File: source/python/bst_levelorder.py
from collections import deque

class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def insert(root, value):
    if root is None:
        return Node(value)
    if value < root.value:
        root.left = insert(root.left, value)
    elif value > root.value:
        root.right = insert(root.right, value)
    return root

def levelorder(root):
    if root is None:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        result.append(node.value)
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)
    return result

root = None
for v in [50, 30, 70, 20, 40, 60, 80]:
    root = insert(root, v)

result = levelorder(root)
print(f"levelorder: {' '.join(str(x) for x in result)}")

File: source/rust/bst_levelorder.rs
use std::collections::VecDeque;

type Tree = Option<Box<Node>>;

struct Node {
    value: i32,
    left: Tree,
    right: Tree,
}

fn insert(root: Tree, value: i32) -> Tree {
    match root {
        None => Some(Box::new(Node { value, left: None, right: None })),
        Some(mut node) => {
            if value < node.value { node.left = insert(node.left, value); }
            else if value > node.value { node.right = insert(node.right, value); }
            Some(node)
        }
    }
}

fn levelorder(root: &Tree) -> Vec<i32> {
    let mut result = Vec::new();
    if root.is_none() { return result; }
    let mut queue: VecDeque<&Node> = VecDeque::new();
    queue.push_back(root.as_ref().unwrap());
    while let Some(node) = queue.pop_front() {
        result.push(node.value);
        if let Some(l) = &node.left { queue.push_back(l); }
        if let Some(r) = &node.right { queue.push_back(r); }
    }
    result
}

fn main() {
    let mut root: Tree = None;
    for &v in &[50, 30, 70, 20, 40, 60, 80] {
        root = insert(root, v);
    }
    let result = levelorder(&root);
    let s: Vec<String> = result.iter().map(|x| x.to_string()).collect();
    println!("levelorder: {}", s.join(" "));
}
